[ADD] estate: Added the Estate Module

snrav-odoo opened the pull request:
Added Estate Module for Managing Real Estate Properties

- This PR adds a new Estate Management Module that allows users to create, manage, buy, and sell real estate  properties. It includes property workflows, offer handling, validations, and data integrity rules.

- To provide a structured and reliable system for managing real estate operations within Odoo, ensuring clean workflows and preventing invalid or inconsistent data through proper constraints and validations.

Changes made :

- Added new estate module with full property management features.
- Created models: estate.property, estate.property.offer, estate.property.type.
- Implemented workflow: new → offer received → offer accepted → sold.
- Added UserError validations for invalid operations.
- Added SQL & Python constraints (price validation, offer price > 0, selling price rules).
- Added views, menus, actions, and access control rules.

snrav-odoo committed:
[IMP] estate: add initial estate application

Created the necessary __init__.py and __manifest__.py files
to initialize the estate application.

snrav-odoo committed:
[IMP] estate: add initial estate module structure

Introduces the basic scaffolding for the new real estate module.
• Creates the module initialization files (__init__).
• Adds the first 'estate.property' model definition.
• Includes the initial set of base fields for the model.

These changes set up the foundational structure of the estate module, enabling
future development of property management features.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1048.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1048)

snrav-odoo committed:
[IMP] estate: Implement access rights for property model

- Added a security CSV file in the `security` folder.
- Granted users access to the `estate.property` model.

These changes ensure that only authorized users can access and manage estate
properties, improving data security and compliance with access control rules.

snrav-odoo committed:
[IMP] estate: add property menus and actions

• Added Real Estate menu structure and linked property action.
• Created window action for `estate.property`.
• Loaded property views and menu XML in manifest.

These changes provide users with a navigation structure and actions to access
property records efficiently, improving usability and making the estate module
functional from the UI.

snrav-odoo committed:
[IMP] estate: Add basic views for property

* Added estate property form view.
* Added estate property list view.
* Added search view with filter and group_by options.

These changes provide the basic UI for managing estate properties, allowing users
to create, view, and search properties efficiently.

snrav-odoo committed:
[IMP] estate: model relations (M2O, O2M, M2M)

* Added estate_property_offer model with required fields.
* Added estate_property_tag model with required fields.
* Added estate_property_type model with required fields.
* Implemented relations between models (Many2one, One2many, Many2many).
* Created corresponding views for offers, tags, and property types.
* Added access rights for new models.

snrav-odoo committed:
[IMP] estate: add computed fields and onchanges for offers and garden.
- Compute total_area and best_price in estate.property
- Compute validity date for offers with inverse method
- Set default garden area and orientation using onchange
- Chapter 8

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,72 @@
+from odoo import models, api, fields
+from dateutil.relativedelta import relativedelta
```
```suggestion
from dateutil.relativedelta import relativedelta

from odoo import models, api, fields
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,72 @@
+        ('north', 'North'),
+        ('south', 'South'),
+        ('east', 'East'),
+        ('west', 'West'),
```
Please follow the same quote throughout the file.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,72 @@
+    tag_ids = fields.Many2many("estate.property.tag", string="Property Tags")
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offer")
+    total_area = fields.Integer(compute="_compute_area")
+    best_price = fields.Integer(compute="_compute_highest")
```
I think it is good practice to declare a compute method name similar to the field name so others can easily identifity it. But it's my personal opinion :sweat_smile:

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,27 @@
+from odoo import models, api, fields
+from dateutil.relativedelta import relativedelta
```
First, we can import external libraries.

bit-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+estate.access_estate_property,"access_estate_property",estate.model_estate_property,base.group_user,1,1,1,1
+estate.access_estate_property_type,"access_estate_property_type",estate.model_estate_property_type,base.group_user,1,1,1,1
+estate.access_estate_property_tag,"access_estate_property_tag",estate.model_estate_property_tag,base.group_user,1,1,1,1
+estate.access_estate_property_offer,"access_estate_property_offer",estate.model_estate_property_offer,base.group_user,1,1,1,1
```
There should be one empty line at the EOF.

bit-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,41 @@
+        </field>
+    </record>
+
+</odoo>
```
There should be one empty line at the EOF.

bit-odoo commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,23 @@
+            </list>
+        </field>
+    </record>
+</odoo>
```
There should be one empty line at the EOF.

bit-odoo commented on estate/views/estate_property_type_views.xml:
```diff
@@ -0,0 +1,24 @@
+            </list>
+        </field>
+    </record>
+</odoo>
```
There should be one empty line at the EOF.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0"?>
+<odoo>
+
+ <record id="estate_property_list_view" model="ir.ui.view">
```
Indentation issues.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,108 @@
+    </record>
+
+
+ <record id="estate_property_form_view" model="ir.ui.view">
```
Indentation issue.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,108 @@
+                                <field name="state"/>
+                        </group>
+                        </group>
+                        </page>
```
Same Indentation issue.

bit-odoo reviewed (commented):
Hello @snrav-odoo 

You did a great job :clap: 

I have added some suggestions.
And can you please improve your commit message, and pr message, and title regarding these guidelines - https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html

Thanks

snrav-odoo committed:
[IMP] estate: add buttons to enforce property and offer workflow

Link business logic to UI buttons.
Prevent cancelled properties from being sold.
Prevent sold properties from being cancelled.
Add ‘Accept’ and ‘Refuse’ buttons for offers.
Update buyer and selling price automatically.
Ensure consistent workflow and reduce errors.

snrav-odoo committed:
[IMP] estate: add SQL and Python constraints

- Added SQL constraints:
  * Expected price must be strictly positive.
  * Selling price must be positive.
  * Offer price must be strictly positive.
  * Property tag name must be unique.
  * Property type name must be unique.

- Added Python constraint:
  * Ensure the selling price cannot be lower than 90% of the expected price
    (except when selling price is zero before offer validation).

snrav-odoo commented:
Hello @bit-odoo sir,

Thanks for the review.
 I have made the changes suggested by you. Kindly have a look into it.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+    )
+    _check_selling_price = models.Constraint(
+        "CHECK(selling_price >= 0)", "Selling price must be positive."
+    )
```
It is good if we can declare it after the field declaration. You can refer odoo codebase.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+from dateutil.relativedelta import relativedelta
+from odoo import models, api, fields, exceptions
+from odoo.exceptions import ValidationError
```
```suggestion
from dateutil.relativedelta import relativedelta

from odoo import models, api, fields, exceptions
from odoo.exceptions import ValidationError
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+        string="Status",
+        default="new",
+    )
+    active = fields.Boolean(default=False)
```
Why default `False`?

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+    )
+    tag_ids = fields.Many2many("estate.property.tag", string="Property Tags")
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offer")
+    total_area = fields.Integer(compute="_total_area")
```
It is good that we can start the compute method with `_compute_fieldname`.So others can easily identify by the name.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+    def action_cancel_offer(self):
+        for record in self:
+            if record.state == "sold":
+                raise exceptions.UserError("Properties which are Sold cannot be Cancelled")
```
It's good to make the error message translatable.
 `raise exceptions.UserError(_("Properties which are Sold cannot be Cancelled"))`
 
 We can use filtered() here.
 
 `if self.filtered(lamda x: x.state=="sold")`

bit-odoo replied:
Please adapt these changes.
https://github.com/odoo/tutorials/pull/1048#discussion_r2609402952

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,105 @@
+            if record.state == "cancelled":
+                raise exceptions.UserError("Properties which are Cancelled cannot be Sold")
+            else:
+                record.state = "sold"
```
It's good to make the error message translatable.
`raise exceptions.UserError(_("Properties which are Cancelled cannot be Sold"))`

We can use filtered() here.

`if self.filtered(lamda x: x.state=="cancelled")`

bit-odoo replied:
Please adapt these changes.

bit-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,43 @@
+        <field name="view_mode">list,form</field>
+      </record>
+
+  <record id="action_estate_offer_view_list" model="ir.ui.view">
```
Indentation issue.Should be one tab.

bit-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,43 @@
+        </field>
+    </record>
+
+  <record id="action_estate_offer_view_form" model="ir.ui.view">
```
Indentation issue.Should be one tab.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,106 @@
+            <field name="living_area" string="Living Area (sqm)" />
+            <field name="expected_price" />
+            <field name="selling_price" />
+            <field name="date_availability" string="Available From" />
```
Indentation issue.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,106 @@
+                  </group>
+               </group>
+               <notebook>
+                  <page string="Description">
```
Indentation issue.should be one tab.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,106 @@
+                  </page>
+                  <page string="Information">
+                     <group>
+                        <field name="salesperson" />
```
Indentation issue.should be one tab.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,106 @@
+      <field name="view_mode">list,form</field>
+      <field name="help" type="html">
+        <p class="o_view_nocontent_smiling_face">Create your Properties here !</p>
+      </field>
```
same.

bit-odoo reviewed (commented):
Hello, Nice progress!
I have made some suggestions.

And can you please update PR title.

snrav-odoo committed:
[IMP] estate: Enhance UI and property management features

• Enable inline editing for property tags and offers without opening form views
• Add widgets and visual decorations to property and offer records
• Introduce a status bar to visualize property states
• Add a stat button on property types to view all related offers

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'data': [
+        'security/ir.model.access.csv',
+        'views/estate_property_views.xml',
+         'views/estate_property_offer_views.xml',
```
indentation issue.

snrav-odoo committed:
[IMP] estate: add inheritance-based business logic
• Restrict property deletion by state
• Update property state on offer creation
• Prevent offers below best price
• Show related properties on users form

snrav-odoo committed:
[ADD] estate_account: create customer invoice when property is sold

Create a link module between estate and account.
Extend estate.property action_sold to generate a customer invoice
with two invoice lines:
• 6% commission on selling price
• Fixed administrative fee of 100.00

snrav-odoo commented:
Hello @bit-odoo sir,

Thanks for the review.
I have resolved the queries raised by you. Kindly have a look into it.

snrav-odoo committed:
[IMP] estate:  Add kanban view for properties using QWeb templates

snrav-odoo committed:
[IMP] awesome_owl: introduce Owl components playground exercises

Add Counter and Card components to demonstrate Owl fundamentals:
- Implement reactive Counter using useState and event handlers
- Extract Counter as reusable sub-component
- Add Card component with props, validation, and markup support
- Demonstrate parent-child communication with callback props
- Compute and display sum of multiple counters in Playground

snrav-odoo committed:
[IMP] awesome_owl: add todo list and advanced card features

- Introduce TodoList and TodoItem components
- Implement todo creation, toggle completion, and deletion
- Add dynamic classes based on todo state
- Use refs and lifecycle hooks to autofocus input
- Add callback props for child-to-parent communication
- Refactor Card component to use slots
- Add toggle button to show or hide card content

snrav-odoo committed:
[IMP] awesome_dashboard: add layout, navigation buttons and dashboard items

- Use the Layout component to provide a standard Odoo screen structure
with a control panel and main content area.

- Add quick navigation buttons to open Customers and Leads using the
action service.

- Introduce a reusable DashboardItem component displaying content in a
card layout with configurable width based on a size prop.

snrav-odoo committed:
[IMP] awesome_dashboard: load statistics, cache rpc and add pie chart

Load real business statistics from /awesome_dashboard/statistics using rpc
and display them as dashboard cards (orders count, amounts, averages, delays).

Introduce a statistics service using memoize to cache network calls and avoid
reloading data each time the dashboard action is opened.

Add a PieChart component with lazy-loaded Chart.js to display the distribution
of sold t-shirts by size (S/M/L/XL/XXL) in the dashboard.

snrav-odoo committed:
[IMP] real_estate: add chatter support to property records
Enable mail.thread and mail.activity.mixin on real estate
properties to track messages, activities, and followers.

snrav-odoo committed:
[IMP] awesome_dashboard: reactive statistics refresh and lazy-loaded dashboard

This commit introduces two improvements to the dashboard module.

1) Periodic refresh of statistics
The statistics service now provides a reactive object instead of a
memoized result. Data is refreshed every 10 minutes (10 seconds during
development) using a timer, and updates are applied in place so any
dashboard components subscribing to it are automatically re-rendered.

This keeps fast-moving data up to date while still benefiting from
caching between loads.

2) Lazy loading of dashboard assets
All dashboard-related code was moved to a /dashboard subfolder and added
to a dedicated asset bundle .

The dashboard component is now registered in the lazy_components
registry and loaded on demand via an intermediate action using
LazyComponent. This reduces the initial page load and only loads the
dashboard when needed.

snrav-odoo committed:
[IMP] awesome_dashboard: make dashboard generic using dashboard items

This change removes hard-coded dashboard content and replaces it with a
generic rendering system based on dashboard item objects. Each item
describes its id, description, component, size and optional props
function, allowing flexible composition of dashboards.

Two reusable item components are introduced:
- NumberCard: displays a title and numeric value
- PieChartCard: displays labels with a pie chart visualization

bit-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_Item/dashboard_item.js:
```diff
@@ -0,0 +1,9 @@
+        size: Number,
+        slots: Object
+    }
+}
```
Should be one empty line at EOF

bit-odoo commented on awesome_dashboard/static/src/dashboard/number_card/number_card.xml:
```diff
@@ -0,0 +1,11 @@
+        </div>
+    </t>
+
+</templates>
```
Should be one empty line at EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart.js:
```diff
@@ -0,0 +1,42 @@
+            }
+        }
+    }
+}
```
Should be one empty line at EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart.xml:
```diff
@@ -0,0 +1,8 @@
+        <canvas t-ref="canvaRef" />
+    </t>
+
+</templates>
```
Should be one empty line at EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart_card.xml:
```diff
@@ -0,0 +1,11 @@
+        </div>
+    </t>
+
+</templates>
```
Should be one empty line at EOF.

bit-odoo commented on awesome_owl/static/src/card/card.js:
```diff
@@ -0,0 +1,19 @@
+    toggle() {
+        this.state.open = !this.state.open;
+    }
+}
```
Should be one empty line at EOF.

bit-odoo commented on awesome_owl/static/src/counter/counter.js:
```diff
@@ -0,0 +1,18 @@
+    decrement() {
+        this.state.value--;
+    }
+}
```
Should be one empty line at EOF.

bit-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,16 @@
+			</button>
+		</div>
+	</t>
+</templates>
```
Should be one empty line at EOF.

bit-odoo commented on awesome_owl/static/src/playground.xml:
```diff
@@ -1,10 +1,28 @@
-<?xml version="1.0" encoding="UTF-8" ?>
+<?xml version="1.0" encoding="UTF-8"?>
```
unneccary diff.

bit-odoo commented on awesome_owl/static/src/todo/todo_list.xml:
```diff
@@ -0,0 +1,21 @@
+			</t>
+		</div>
+	</t>
+</templates>
```
Should be one empty line at EOF.

bit-odoo reviewed (commented):
Hello,
i have made some comments.
Thanks

snrav-odoo committed:
[IMP] estate: auto-reject other offers when one is accepted
When an offer is accepted on a property, all remaining offers are now
automatically set to 'rejected'.

snrav-odoo commented:
Hello @bit-odoo sir,

I have resolved the queries suggested by you. Kindly have a look into it. 

Thanks

snrav-odoo committed:
[IMP] awesome_owl: support multiple counters and correct sum updates

- Added decrement buttons for both counters.
- Updated logic so the global sum recalculates correctly on increment/decrement.
- Improved UI consistency across counters.

kcv-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -11,3 +12,16 @@ class EstatePropertyOffer(models.Model):
+    @api.depends("validity")
+    def _compute_deadline(self):
+        for record in self:
+            base_date = record.create_date or fields.Date.today()
```
In which case `record.create_date` is not set?

snrav-odoo replied:
record.create_date is not set when the record has not yet been persisted to the database. This typically occurs:

1) When the record is new and unsaved (e.g., while creating a record in the form view before clicking Save)
2) During compute or onchange execution on a virtual record

kcv-odoo commented on estate/models/estate_property.py:
```diff
@@ -46,3 +46,27 @@ class EstateProperty(models.Model):
+    @api.depends("offer_ids.price")
+    def _compute_highest(self):
+        for record in self:
+            if not record.mapped("offer_ids.price"):
+                record.best_price = 0
+            else:
+                record.best_price = max(record.mapped("offer_ids.price"))
```
How does mapped work? Isn't it better to use variable and do mapped only once?

Could you use `_read_group` instead accessing x2Many field in a loop? And do you why `_read_group` is better then this?

kcv-odoo commented on estate/models/estate_property.py:
```diff
@@ -46,3 +46,27 @@ class EstateProperty(models.Model):
+            self.garden_orientation = "north"
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
```
Could you convert this to compute method? What are the drawback of the on change?

snrav-odoo replied:
Yes we can convert it. I have converted this to a compute method in my commit. 

Also below mentioned are the drawback of the on change method. 

1) It is triggered only in the form view (not in backend operations like: scheduled job).
2) Data set through on change cannot be reliably used for filtering or grouping if not consistently stored.
3) Fields updated via on change are not persisted unless the record is explicitly saved, which may result in missing values.

kcv-odoo commented on estate/models/estate_property.py:
```diff
@@ -70,3 +70,19 @@ def _onchange_garden(self):
+    def action_sold_property(self):
+        for record in self:
+            if record.state == "cancelled":
+                raise exceptions.UserError("Properties which are Cancelled cannot be Sold")
```
Do you know about translations?

snrav-odoo replied:
Yes, I am aware of translations in Odoo.

All user-facing strings should be wrapped in the _() function from odoo to make them translatable.

I have added the translations wherever it were missing in my commit.

kcv-odoo commented on estate/models/estate_property.py:
```diff
@@ -86,3 +92,14 @@ def action_cancel_offer(self):
                 record.state = "cancelled"
         return True
+
+    @api.constrains("selling_price", "expected_price")
+    def _check_selling_price_percentage_criteria(self):
```
What is difference b/w this constraints and SQL constraints?

snrav-odoo replied:
Differences Between Python @api.constrains and SQL _sql_constraints

1) Complexity:

- Python @api.constrains: Can include complex logic
- SQL _sql_constraints: Limited to simple conditions

2) Error type:

- Python @api.constrains: Raises ValidationError
- SQL _sql_constraints: Database error

3) Performance:

- Python @api.constrains: Slightly slower due to ORM 
- SQL _sql_constraints: Very fast, handled directly by the database

kcv-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -1,106 +1,109 @@
+		<field name="arch" type="xml">
+			<form string="Property">
+				<header>
+					<button name="action_sold_property" type="object" string="Sold" class="oe_highlight" invisible="state in ('sold','cancelled','new','offer_received')"/>
```
What is the difference b/w type object and type action?

snrav-odoo replied:
The major difference in using type="object" is that it calls a Python method on the record (server-side logic) like we have called action_sold_property function, while type="action" triggers a predefined Odoo client/server action (like opening a report, wizard, or view). 

Hence generally we use object for business logic and action for navigation or external actions.

kcv-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -1,22 +1,27 @@
     )
     partner_id = fields.Many2one("res.partner", string="Partner", required=True)
     property_id = fields.Many2one("estate.property", string="Property", required=True)
+    property_type_id = fields.Many2one(related="property_id.property_type_id", store=True)
```
Why do we need to store this related field?

snrav-odoo replied:
store=True on the property_type_id field is used to persist its value in the database.

It will allow searching, filtering, grouping, and improves performance when accessing the field in views or reports.

If we don't write store=True then it will be computed every time which will increase the time complexity.

kcv-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -1,10 +1,22 @@
+    @api.depends("offer_ids")
+    def _compute_offer_count(self):
+        for record in self:
+            record.offer_count = len(record.property_ids.mapped('offer_ids'))
```
Use _read_group here as this is not good for performance

snrav-odoo replied:
I have used _read_group which is added in my commit.

kcv-odoo commented on estate/views/estate_property_type_views.xml:
```diff
@@ -1,25 +1,52 @@
+			<form name="List">
+				<header>
+					<button class="oe_stat_button" type="action" name="estate.action_estate_property_offer" icon="fa-list" string="offer"/>
+					<field name="offer_count"/>
```
Where do we want to display this count in UI?

snrav-odoo replied:
We display this count on the property type screen when we click on individual property type.

kcv-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -1,106 +1,109 @@
+		<field name="name">estate.property.list</field>
+		<field name="model">estate.property</field>
+		<field name="arch" type="xml">
+			<list string="Properties" decoration-success="state in ('offer_accepted')" decoration-bf="state in ('offer_accepted')" decoration-muted="state in ('sold')">
```
What is `decoration-bf` 🤔

snrav-odoo replied:
It stands for bold font.

Green highlight (decoration-success) and bold font (decoration-bf) for offer_accepted properties.

It will make the font bold.

kcv-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -47,3 +48,16 @@ def action_refuse(self):
+
+    @api.model
+    def create(self, vals):
+        if len(vals) > 0:
```
When does vals will not have any length? 🤔

snrav-odoo replied:
This check is not necessary so I have removed it.

kcv-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -47,3 +48,16 @@ def action_refuse(self):
+                raise UserError(
+                    "Offer with an amount lower than an existing offer cannot be created."
+                )
+        return super().create(vals)
```
This seems so bad. can you optimize it?

snrav-odoo replied:
I have tried to optimize it in my commit.

kcv-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,30 @@
+
+class EstateProperty(models.Model):
+    _inherit = "estate.property"
+
+    def action_sold_property(self):
+        self.env["account.move"].create(
+            {
+                "partner_id": self.customer.id,
+                "move_type": "out_invoice",
+                "invoice_line_ids": [
+                    Command.create(
+                        {
+                            "name": self.name,
+                            "quantity": 1,
+                            "price_unit": self.selling_price * 0.6,
+                    }
+                    ),
+                    Command.create(
+                        {
+                            "name": "Administrative fees",
+                            "quantity": 1,
+                            "price_unit": 100,
+                        }
+                    ),
+                ],
+            }
+        )
+        return super().action_sold_property()
```
What will happen if property is already cancelled? 💭

snrav-odoo replied:
Added a check in `action_sold_property` to prevent creating invoices for properties that are cancelled.

Although the Sold button is hidden in the UI for cancelled properties, this ensures that the method cannot be triggered via API calls or other modules, maintaining data integrity and preventing incorrect invoice creation.

I have added code changes in my commit.

kcv-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,30 @@
+                        {
+                            "name": "Administrative fees",
+                            "quantity": 1,
+                            "price_unit": 100,
```
Could you make fees dynamic depending on salesperson? 🙂

snrav-odoo replied:
Yes I have made fees dynamic depending on the salesperson you can check my commit.

kcv-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,30 @@
+                        {
+                            "name": self.name,
+                            "quantity": 1,
+                            "price_unit": self.selling_price * 0.6,
```
What are we doing here? 🤔

snrav-odoo replied:
Here we are over ridding action_sold_property function to automatically create a customer invoice when a property is sold.

The invoice includes two lines:

1) 6% of the selling price as the main property fee.
2) 100.00 as administrative fees.

kcv-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,30 @@
+    _inherit = "estate.property"
+
+    def action_sold_property(self):
+        self.env["account.move"].create(
```
How does invoice linked to property? 🤔

snrav-odoo replied:
The invoice created during action_sold_property is linked to the property indirectly through the invoice lines, using the property name and selling price.

kcv-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_Item/dashboard_item.xml:
```diff
@@ -0,0 +1,15 @@
+            t-att-style="'width: ' + width"
+        >
+            <div class="card-body">
+                <t t-slot="default"/>
```
How does t-slot works?

snrav-odoo replied:
Introduced t-slot in the dashboard_item template to create a reusable dashboard card component.
The slot allows dynamic content to be injected from the main dashboard while keeping layout and styling centralized.
This improves code reuse, readability, and makes the dashboard easily extensible.

kcv-odoo commented on awesome_dashboard/static/src/dashboard_action.js:
```diff
@@ -0,0 +1,16 @@
+    static components = { LazyComponent };
+
+    static template = xml`
+        <LazyComponent 
```
What is LazyComponent? 🤔

snrav-odoo replied:
Implemented LazyComponent to load the Awesome Dashboard assets only when the dashboard action is triggered.
This improves performance by reducing initial bundle size and ensures the dashboard module is loaded on demand.

kcv-odoo reviewed (changes requested):
Thank you for your work!

Added some question regarding overall odoo codebase, try to explain in simpler way! and keep the separate commit for review change

Have a nice day!

snrav-odoo committed:
[IMP] estate: optimize offer logic, computed fields, and admin fees

- Converted 'garden' field from onchange to compute method for better reactivity
- Optimized create() method of estate.property.offer to handle offers efficiently.
- Replaced mapped() with read_group() in computing offer counts to improve performance.
- Made administrative fees dynamic based on salesperson

snrav-odoo committed:
[IMP] estate: add administrative fees dynamic based on salesperson

snrav-odoo committed:
[IMP] estate: add administrative fees dynamic based on salesperson

snrav-odoo committed:
[IMP] sale: % Delivered field and Delivered label update

Before:
- Sale order lines only displayed Delivered quantity.
- Delivered value was not editable for services invoiced based on
  delivered quantity (manual).
- No percentage-based delivery tracking was available.

After:
- Added a % Delivered field on sale order lines.
- Delivered column label updated to reflect percentage-based delivery.
- % Delivered is editable only for services with manual delivered
  quantity invoicing and syncs with delivered quantity.

Impact:
- Enables percentage-based delivery tracking for manual service
  invoicing.
- Improves usability and accuracy of delivered service billing.

task- 5387190

snrav-odoo committed:
[IMP] sale: % Delivered field and Delivered label update
Before:
- Sale order lines only displayed Delivered quantity.
- Delivered value was not editable for services invoiced based on
  delivered quantity (manual).
- No percentage-based delivery tracking was available.

After:
- Added a % Delivered field on sale order lines.
- Delivered column label updated to reflect percentage-based delivery.
- % Delivered is editable only for services with manual delivered
  quantity invoicing and syncs with delivered quantity.

Impact:
- Enables percentage-based delivery tracking for manual service
  invoicing.
- Improves usability and accuracy of delivered service billing.

task- 5387190
